Customizable album decorations

evoludolab opened the pull request:
This pull request that complements LycheeOrg/Lychee-front#364 to display the number of subalbums and of photos on the album cover (more details there) and takes into account the helpful feedback on #1629.

evoludolab committed:
optional display of subalbum and photo counts

- new settings 'show_num_albums' and 'show_num_photos' added
- album.blade should produce the same html as the frontend. not tested.

evoludolab committed:
make phpstan happy

ildyria commented:
I like the idea, but that is not the correct execution. I would try to relation rather than having a single query per album. The problem currently is that you are having a `n+1` query problem. If you have 5 albums, it is fast, if you have 100 albums... that is another story...

PS: I already tried to use `Model::$withCount` , it triggers and infinite loop.

I would try to have a look at the cover Thumb relation in albums.

evoludolab commented:
Thanks for the feedback! Bummer, I feared that this is going to be the case... and the reason I was happy to discover the shortcut used in #1629 😉

Unfortunately, I know close to nothing about databases, queries and relations but it's interesting and I am motivated to learn all about it. Any suggestions are most welcome but for now I have a start with the cover thumb relation, thanks! Since this might take some time, I converted the PR into a draft.

evoludolab committed:
Merge remote-tracking branch 'upstream/master' into num_albums_photos

evoludolab committed:
remove spurious copy & paste markers

evoludolab committed:
Merge remote-tracking branch 'upstream/master' into num_albums_photos

evoludolab commented:
@ildyria Now I could use some more pointers... I hope you don't mind! With patience I managed to put together the SQL query needed to retrieve the albums in the root directory together with their sub-album counts (for easier readability I used capitals for the newly added SQL instructions):
<pre><code>WITH root_albums AS (
	<i>select "albums".* from "albums" inner join "base_albums" on "base_albums"."id" = "albums"."id" 
	left join "user_base_album" on "user_base_album"."base_album_id" = "base_albums"."id" 
		and "user_base_album"."user_id" = ? where "parent_id" is null</i>
)
SELECT albums.*, count(*)-1 AS num_albums
FROM root_albums
LEFT JOIN albums ON (albums.id = root_albums.id OR albums.parent_id = root_albums.id)
GROUP BY root_albums.id
</pre></code>
which adds a new column `num_albums` with the sub-album count. The subquery for `root_albums` is typeset in italics and represents the `$query` string currently used by Lychee (see Top.php:82) and will vary depending on the user etc. It's just a copy of what `$query->toSql()` returns in my developer environment.

Is this approach feasible? If so, any suggestions for converting the query into the Laravel framework would be most appreciated. In particular, dealing with the subquery is currently a closed book... The extension to do the same for all other albums should then be fairly straight forward, I assume.

**Edit:** code *italicized* for clarity and explanations extended.

ildyria commented:
See https://github.com/LycheeOrg/Lychee/tree/countSubAlbums , more precisely https://github.com/LycheeOrg/Lychee/commit/ef39b2251ad4542212fddce3a7da7618bf7e9984

evoludolab commented:
Wow, what a response time! Thank you very much!

Now I have plenty to play with over the holidays! My plan is to
- first understand your implementation!!! 
- copy your code to my WIP branch
- enable customizable setting for subalbum counts (for some caveats concerning this, see LycheeOrg/Lychee-front#364)
- apply same logic to retrieve the number of photos in `num_photos`
- hit the 'ready for review' button 😄

evoludolab committed:
sub-albums count returned in query

suggestion by @ildyria see https://github.com/LycheeOrg/Lychee/tree/countSubAlbums
this only adds configuration settings and uses 'num_subalbums' instead of 'num_children'

evoludolab committed:
photos count returned in query

adapted from strategy to count sub-albums and inspired by @ildyria

evoludolab committed:
change settings for album decorations

- replace 'show_num_albums' and 'show_num_photos' by 'album_decorations', which can be one of 'none|original|album|photo|all'
- add 'album_decoration_orientation' to specify how to arrange decorations; only matters with 'album_decorations == all'

ildyria commented on app/Models/Album.php:
```diff
@@ -27,6 +27,7 @@
  * @property string|null       $parent_id
  * @property Album|null        $parent
  * @property Collection<Album> $children
+ * @property int               $num_subalbums    The number of children.
```
I would use `$num_children` to be consistent with the `$children` attribute.

evoludolab replied:
The key `children` gets later renamed to `albums` (l.186) that's why I went with albums to begin with. However, I like your emphasis on consistency and now followed your suggestion. Even though I also prefer the term children, upon return I nevertheless rename the attribute to `num_subalbums` also for consistency. In any case, I have no strong preferences either way but at least using `num_albums` I find a bit confusing because it is not immediately clear whether it refers to this album (the parent, i.e. its siblings) or its children.

ildyria commented on app/Models/Extensions/AlbumBuilder.php:
```diff
@@ -87,10 +93,22 @@ public function getModels($columns = ['*']): array
-				'min_taken_at' => $minTsSelect,
-				'max_taken_at' => $maxTsSelect,
-			]);
+			if (Configs::getValueAsBool('show_num_albums')) {
```
No, this would make the attributes of the object change depending of the config. This significantly complexify the code both for future and for testing.

Just have this as a config sent to the front-end and act on it in the front-end, not in the back-end. It is not hiding something sensitive as the user actually see the number of sub albums and would also be able to open said album and count.

evoludolab replied:
This should not have been there and has since been fixed by changing the settings from `show_num_albums` and `show_num_photos` to the more general and IMHO preferred `album_decorations` setting. However, I still process the settings in the back-end as an attempt to simplify the queries as much as possible. The attributes `num_subalbums` and `num_photos` are now always returned but just set to `-1` if no query was issued.

ildyria commented on resources/views/livewire/parts/album.blade.php:
```diff
@@ -45,9 +45,15 @@
-@if ((isset($data['has_albums']) && $data['has_albums'] == '1') || (isset($data['albums']) && count($data['albums']) > 0))
-<div class='subalbum_badge'>
-	<x-icon class='badge--folder' icon='layers' />
+<div class='counters'>
```
you can ignore this file, it is likely to be nuked in the future.

ildyria commented on app/Models/Extensions/AlbumBuilder.php:
```diff
@@ -87,10 +93,57 @@ public function getModels($columns = ['*']): array
+			$showSubalbum = true;
+			$showSubalbumCount = false;
+			$showPhotoCount = false;
+			switch (Configs::getValueAsString('album_decorations')) {
```
See comments in commits.

evoludolab replied:
In order to keep the complexity of queries to a minimum, the back-end considers the settings but now always returns the same set of attributes, i.e. `num_subalbums` and `num_photos`. Does this address your concerns?

evoludolab commented:
@ildyria Just a heads-up the photo count is currently buggy and I have a number of changes waiting to be pushed. Thanks for your feedback so far but otherwise it's probably too early for any serious review... Anyways, I will get to your comments above soon.

evoludolab committed:
fix visibility for photos

- streamline visibility queries
- make phpstan happier

evoludolab committed:
addressing comments by @ildyria

- rename 'num_subalbums' to 'num_children' for consistency
- returning attribute 'num_subalbums' b/c the front-end expects 'albums' instead of 'children' (and 'num_albums' seems a tad ambiguous)
- ensure that always the same set of attributes is returned to the front-end
- improve query for `num_children`

evoludolab commented:
@ildyria now the PR is getting close to be ready for review. I have pushed the changes that were needed for the photos count and addressed your comments as best I can.

What is left is that I would like to add the settings for the `album_decorations` and their `album_decoration_orientation` to the main settings page with a pull-down lists. Otherwise these settings seem way too obscure... Besides, changing the decorations needs to trigger an update of the album to reflect them. Even though all this should not affect the back-end, I'll leave the draft stage until I am satisfied with both parts 😄

evoludolab committed:
settings for album decorations added

- route to set decorations added
- localizations
- query for 'original' decorations fixed

evoludolab committed:
fix labels of pull down menu in settings

further minor changes to labels

evoludolab committed:
Merge remote-tracking branch 'upstream/master' into num_albums_photos

evoludolab committed:
fix merge conflicts and refactoring

evoludolab committed:
Merge remote-tracking branch 'upstream/master' into num_albums_photos

evoludolab committed:
more fixes in response to refactoring

evoludolab committed:
Merge remote-tracking branch 'upstream/master' into num_albums_photos

ildyria commented on app/Http/Requests/Settings/SetAlbumDecorationRequest.php:
```diff
@@ -0,0 +1,63 @@
+
+class SetAlbumDecorationRequest extends BaseApiRequest
+{
+	public const ALBUM_DECORATION_ATTRIBUTE = 'album_decoration';
```
Use `App\Contracts\Http\Requests\RequestAttribute` instead.

ildyria commented on app/Http/Requests/Settings/SetAlbumDecorationRequest.php:
```diff
@@ -0,0 +1,63 @@
+class SetAlbumDecorationRequest extends BaseApiRequest
+{
+	public const ALBUM_DECORATION_ATTRIBUTE = 'album_decoration';
+	public const ALBUM_DECORATION_ORIENTATION_ATTRIBUTE = 'album_decoration_orientation';
```
Use `App\Contracts\Http\Requests\RequestAttribute` instead.

ildyria commented on app/Locale/French.php:
```diff
@@ -51,6 +51,7 @@ public function get_locale(): array
 			'DEFAULT_LICENSE' => 'License par defaut pour les nouveaux ajouts :',
 			'SET_LICENSE' => 'Sélectioner une license',
 			'SET_OVERLAY_TYPE' => 'Sélectioner le type d’Overlay',
```
```suggestion
			'SET_OVERLAY_TYPE' => 'Sélectionner le type d’Overlay',
```
typo ;p

evoludolab replied:
There were a few more of the same type 😉

ildyria commented on app/Locale/French.php:
```diff
@@ -51,6 +51,7 @@ public function get_locale(): array
 			'DEFAULT_LICENSE' => 'License par defaut pour les nouveaux ajouts :',
 			'SET_LICENSE' => 'Sélectioner une license',
 			'SET_OVERLAY_TYPE' => 'Sélectioner le type d’Overlay',
+			'SET_ALBUM_DECORATION' => 'Set album decorations',
```
```suggestion
			'SET_ALBUM_DECORATION' => 'Sélectionner les décorations des album',
```

ildyria commented on app/Locale/French.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
 			'OVERLAY_EXIF' => 'Informations EXIF',
 			'OVERLAY_DESCRIPTION' => 'Description de la photo',
 			'OVERLAY_DATE' => 'Date de la photo',
+			'ALBUM_DECORATION' => 'Album decorations:',
```
```suggestion
			'ALBUM_DECORATION' => 'Décorations des album:',
```

ildyria commented on app/Locale/French.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
 			'OVERLAY_DESCRIPTION' => 'Description de la photo',
 			'OVERLAY_DATE' => 'Date de la photo',
+			'ALBUM_DECORATION' => 'Album decorations:',
+			'ALBUM_DECORATION_NONE' => 'No badges',
```
```suggestion
			'ALBUM_DECORATION_NONE' => 'Pas de décorations',
```

ildyria commented on app/Locale/French.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
 			'OVERLAY_DATE' => 'Date de la photo',
+			'ALBUM_DECORATION' => 'Album decorations:',
+			'ALBUM_DECORATION_NONE' => 'No badges',
+			'ALBUM_DECORATION_ORIGINAL' => 'Sub-album badge, no count',
```
```suggestion
			'ALBUM_DECORATION_ORIGINAL' => 'Montrer la présence de sous-album',
```

ildyria commented on app/Locale/French.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
+			'ALBUM_DECORATION' => 'Album decorations:',
+			'ALBUM_DECORATION_NONE' => 'No badges',
+			'ALBUM_DECORATION_ORIGINAL' => 'Sub-album badge, no count',
+			'ALBUM_DECORATION_ALBUM' => 'Sub-album badge with count',
```
```suggestion
			'ALBUM_DECORATION_ALBUM' => 'Montrer le nombre de sous-album',
```

ildyria commented on app/Locale/French.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
+			'ALBUM_DECORATION_NONE' => 'No badges',
+			'ALBUM_DECORATION_ORIGINAL' => 'Sub-album badge, no count',
+			'ALBUM_DECORATION_ALBUM' => 'Sub-album badge with count',
+			'ALBUM_DECORATION_PHOTO' => 'Photo badge with count',
```
```suggestion
			'ALBUM_DECORATION_PHOTO' => 'Montrer le nombre de photos',
```

ildyria commented on app/Locale/French.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
+			'ALBUM_DECORATION_ORIGINAL' => 'Sub-album badge, no count',
+			'ALBUM_DECORATION_ALBUM' => 'Sub-album badge with count',
+			'ALBUM_DECORATION_PHOTO' => 'Photo badge with count',
+			'ALBUM_DECORATION_ALL' => 'Sub-album and photo badges with counts',
```
```suggestion
			'ALBUM_DECORATION_ALL' => 'Montrer le nombre de sous-album et de photos',
```

ildyria commented on app/Locale/French.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
+			'ALBUM_DECORATION_ALBUM' => 'Sub-album badge with count',
+			'ALBUM_DECORATION_PHOTO' => 'Photo badge with count',
+			'ALBUM_DECORATION_ALL' => 'Sub-album and photo badges with counts',
+			'ALBUM_DECORATION_ORIENTATION' => 'Orientation of album decorations:',
```
```suggestion
			'ALBUM_DECORATION_ORIENTATION' => 'Orientation des décorations',
```

ildyria commented on app/Locale/French.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
+			'ALBUM_DECORATION_PHOTO' => 'Photo badge with count',
+			'ALBUM_DECORATION_ALL' => 'Sub-album and photo badges with counts',
+			'ALBUM_DECORATION_ORIENTATION' => 'Orientation of album decorations:',
+			'ALBUM_DECORATION_ORIENTATION_ROW' => 'Horizontal decorations (photos, albums)',
```
```suggestion
			'ALBUM_DECORATION_ORIENTATION_ROW' => 'Horizontale (photos, albums)',
```

ildyria commented on app/Locale/French.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
+			'ALBUM_DECORATION_ALL' => 'Sub-album and photo badges with counts',
+			'ALBUM_DECORATION_ORIENTATION' => 'Orientation of album decorations:',
+			'ALBUM_DECORATION_ORIENTATION_ROW' => 'Horizontal decorations (photos, albums)',
+			'ALBUM_DECORATION_ORIENTATION_ROW_REVERSE' => 'Horizontal decorations (albums, photos)',
```
```suggestion
			'ALBUM_DECORATION_ORIENTATION_ROW_REVERSE' => 'Horizontale (albums, photos)',
```

ildyria commented on app/Locale/French.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
+			'ALBUM_DECORATION_ORIENTATION' => 'Orientation of album decorations:',
+			'ALBUM_DECORATION_ORIENTATION_ROW' => 'Horizontal decorations (photos, albums)',
+			'ALBUM_DECORATION_ORIENTATION_ROW_REVERSE' => 'Horizontal decorations (albums, photos)',
+			'ALBUM_DECORATION_ORIENTATION_COLUMN' => 'Vertical decorations (top photos, albums)',
```
```suggestion
			'ALBUM_DECORATION_ORIENTATION_COLUMN' => 'Verticale (photos au-dessus, albums)',
```

ildyria commented on app/Locale/French.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
+			'ALBUM_DECORATION_ORIENTATION_ROW' => 'Horizontal decorations (photos, albums)',
+			'ALBUM_DECORATION_ORIENTATION_ROW_REVERSE' => 'Horizontal decorations (albums, photos)',
+			'ALBUM_DECORATION_ORIENTATION_COLUMN' => 'Vertical decorations (top photos, albums)',
+			'ALBUM_DECORATION_ORIENTATION_COLUMN_REVERSE' => 'Vertical decorations (top albums, photos)',
```
```suggestion
			'ALBUM_DECORATION_ORIENTATION_COLUMN_REVERSE' => 'Verticale (albums au-dessus, photos)',
```

ildyria commented on app/Models/Album.php:
```diff
@@ -27,7 +27,9 @@
  * @property string|null       $parent_id
  * @property Album|null        $parent
  * @property Collection<Album> $children
+ * @property int               $num_children     The number of children (in original mode this is simply 0 or 1).
```
```suggestion
 * @property int               $num_children     The number of children
```

ildyria commented on app/Models/Extensions/AlbumBuilder.php:
```diff
@@ -87,10 +91,63 @@ public function getModels($columns = ['*']): array
+			$showPhotoCount = false;
+			$isAdmin = false;
+			$userID = null;
+			switch (Configs::getValueAsString('album_decoration')) {
```
This is too complex, do not do logic here.
Simply retrieve the count (but with visibility conditions).

For example this count can be used later for statistic pages etc... As a result, making it conditional is counter productive!

ildyria replied:
Do note that there is no secret in those, as the data is already available when you open the album.
Simply apply the cases in the front-end depending of the configuration selected, not in the back-end.

evoludolab replied:
I guess I was getting paranoid with queries and performance issues.

ildyria reviewed (changes requested):
See comments.

qwerty287 commented on app/Locale/German.php:
```diff
@@ -51,6 +51,7 @@ public function get_locale(): array
 			'DEFAULT_LICENSE' => 'Standard-Lizenz für neue Uploads:',
 			'SET_LICENSE' => 'Lizenz anwenden',
 			'SET_OVERLAY_TYPE' => 'Setze Overlay',
+			'SET_ALBUM_DECORATION' => 'Set album decorations',
```
```suggestion
			'SET_ALBUM_DECORATION' => 'Album-Dekorationen speichern',
```

qwerty287 commented on app/Locale/German.php:
```diff
@@ -383,6 +384,17 @@ public function get_locale(): array
+			'ALBUM_DECORATION_ORIENTATION_ROW' => 'Horizontal decorations (photos, albums)',
+			'ALBUM_DECORATION_ORIENTATION_ROW_REVERSE' => 'Horizontal decorations (albums, photos)',
+			'ALBUM_DECORATION_ORIENTATION_COLUMN' => 'Vertical decorations (top photos, albums)',
+			'ALBUM_DECORATION_ORIENTATION_COLUMN_REVERSE' => 'Vertical decorations (top albums, photos)',
```
```suggestion
			'ALBUM_DECORATION' => 'Album-Dekorationen:',
			'ALBUM_DECORATION_NONE' => 'Keine Abzeichen',
			'ALBUM_DECORATION_ORIGINAL' => 'Abzeichen für Subalbums, kein Zähler',
			'ALBUM_DECORATION_ALBUM' => 'Abzeichen für Subalbums mit Zähler',
			'ALBUM_DECORATION_PHOTO' => 'Abzeichen für Fotos mit Zähler',
			'ALBUM_DECORATION_ALL' => 'Abzeichen für Subalbums und Fotos mit Zähler',
			'ALBUM_DECORATION_ORIENTATION' => 'Ausrichtung der Album-Dekorationen:',
			'ALBUM_DECORATION_ORIENTATION_ROW' => 'Horizontale Dekorationen (Fotos, Alben)',
			'ALBUM_DECORATION_ORIENTATION_ROW_REVERSE' => 'Horizontale Dekorationen (Alben, Fotos)',
			'ALBUM_DECORATION_ORIENTATION_COLUMN' => 'Vertikale Dekorationen (oben Fotos, Alben)',
			'ALBUM_DECORATION_ORIENTATION_COLUMN_REVERSE' => 'Vertikale Dekorationen (oben Alben, Fotos)',
```

evoludolab replied:
Inspired by @ildyria shorter labels in French, I followed the same pattern for German and English.

qwerty287 replied:
Thanks, your translations seem fine.

ildyria commented:
Please do not mark issues are resolved, especially when no commits have been added. :)

evoludolab commented:
Undone. Was just for me to keep track of the many straight forward language related changes. Commit follows swiftly 😉

evoludolab committed:
updates to labels in different languages

inspired by @ildyria 's more concise French labels I followed the same pattern for English and German ( @qwerty287 I hope you approve). The English template for all other languages is updated as well.

evoludolab committed:
use RequestAttribute, simplify queries

evoludolab committed:
Merge remote-tracking branch 'upstream/master' into num_albums_photos

ildyria commented on app/Http/Requests/Settings/SetAlbumDecorationRequest.php:
```diff
@@ -0,0 +1,61 @@
+		return [RequestAttribute::ALBUM_DECORATION_ATTRIBUTE => [
+			'required',
+			'string',
+			Rule::in(['none', 'original', 'album', 'photo', 'all']),
```
Use an Enum instead. See `ColumnSortingType.php` for example and `SetSortingSettingsRequest.php`

ildyria commented on app/Models/Extensions/AlbumBuilder.php:
```diff
@@ -87,9 +90,23 @@ public function getModels($columns = ['*']): array
 			$maxTsSelect = $minTsSelect->clone()->select(DB::raw('MAX(taken_at)'));
 
+			$isAdmin = (Auth::user()?->may_administrate === true);
+			$userID = !$isAdmin ? Auth::id() : null;
```
```suggestion
			$userID = Auth::id();
```
Don't try to do unnecessary optimizations :)

evoludolab committed:
use enums for album decorations

evoludolab committed:
one got away... last of the requested changes

ildyria commented on app/Enum/AlbumDecorationType.php:
```diff
@@ -0,0 +1,27 @@
+	 *
+	 * @return AlbumDecorationType
+	 */
+	public function toAlbumDecorationType(): AlbumDecorationType
```
where is this used ?

ildyria commented on app/Enum/AlbumDecorationOrientation.php:
```diff
@@ -0,0 +1,26 @@
+	 *
+	 * @return AlbumDecorationOrientation
+	 */
+	public function toAlbumDecorationOrientation(): AlbumDecorationOrientation
```
where is this used ?

evoludolab committed:
remove unused methods

evoludolab committed:
Merge remote-tracking branch 'upstream/master' into num_albums_photos

evoludolab committed:
attempt to appease phpunit

ildyria committed:
fix front

ildyria committed:
stronger types

ildyria committed:
sync front

ildyria reviewed (approved):
LGTM, tested.
